File: apps/web/lib/api/history.ts
import { getCached, invalidateCached, prefetchCached } from "../client-cache";
import { invalidateInsightsCache } from "./insights";

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface FeedbackSummary {
  feedback_label: string;
  reason: string | null;
  created_at: string;
}

export interface HistoryItemResponse {
  id: string;
  source: "manual" | "gmail";
  subject: string | null;
  sender: string | null;
  final_prediction: "spam" | "not_spam";
  final_risk_score: number;
  risk_band: "low" | "medium" | "high";
  personalized: boolean;
  saved_at: string;
}

export interface HistoryDetailResponse extends HistoryItemResponse {
  review_state: string | null;
  personalization_reasons: string[] | null;
  agreement_ratio: number;
  model_version: string;
  feedback: FeedbackSummary[];
}

export interface HistoryListResponse {
  items: HistoryItemResponse[];
  next_cursor: string | null;
  total_count: number | null;
}

export interface ClearHistoryResponse {
  deleted_count: number;
}

export interface HistoryQueryParams {
  cursor?: string;
  limit?: number;
  source?: "manual" | "gmail";
  verdict?: "spam" | "not_spam" | "review";
  query?: string;
}

const HISTORY_LIST_CACHE_PREFIX = "history:list:";
const HISTORY_DETAIL_CACHE_PREFIX = "history:detail:";
const HISTORY_LIST_CACHE_TTL_MS = 60 * 1000;
const HISTORY_DETAIL_CACHE_TTL_MS = 5 * 60 * 1000;

function getHistoryListCacheKey(params: HistoryQueryParams): string {
  return `${HISTORY_LIST_CACHE_PREFIX}${JSON.stringify({
    cursor: params.cursor ?? null,
    limit: params.limit ?? null,
    source: params.source ?? null,
    verdict: params.verdict ?? null,
    query: params.query ?? null,
  })}`;
}

export async function getHistory(
  params: HistoryQueryParams = {}
): Promise<HistoryListResponse> {
  const load = async () => {
    const url = new URL(`${API_BASE}/api/v1/history`);
    if (params.cursor) url.searchParams.set("cursor", params.cursor);
    if (params.limit) url.searchParams.set("limit", String(params.limit));
    if (params.source) url.searchParams.set("source", params.source);
    if (params.verdict) url.searchParams.set("verdict", params.verdict);
    if (params.query) url.searchParams.set("q", params.query);

    const res = await fetch(url.toString(), { credentials: "include" });
    if (!res.ok) throw new Error(`Failed to fetch history (${res.status})`);
    return res.json();
  };

  if (params.cursor) {
    return load();
  }

  return getCached({
    key: getHistoryListCacheKey(params),
    ttlMs: HISTORY_LIST_CACHE_TTL_MS,
    loader: load,
  });
}

export async function getHistoryItem(id: string): Promise<HistoryDetailResponse> {
  return getCached({
    key: `${HISTORY_DETAIL_CACHE_PREFIX}${id}`,
    ttlMs: HISTORY_DETAIL_CACHE_TTL_MS,
    loader: async () => {
      const res = await fetch(`${API_BASE}/api/v1/history/${id}`, {
        credentials: "include",
      });
      if (!res.ok) throw new Error(`Failed to fetch history item (${res.status})`);
      return res.json();
    },
  });
}

export function prefetchHistory(params: HistoryQueryParams = {}): void {
  if (params.cursor) return;

  prefetchCached({
    key: getHistoryListCacheKey(params),
    ttlMs: HISTORY_LIST_CACHE_TTL_MS,
    loader: async () => {
      const url = new URL(`${API_BASE}/api/v1/history`);
      if (params.limit) url.searchParams.set("limit", String(params.limit));
      if (params.source) url.searchParams.set("source", params.source);
      if (params.verdict) url.searchParams.set("verdict", params.verdict);
      if (params.query) url.searchParams.set("q", params.query);

      const res = await fetch(url.toString(), { credentials: "include" });
      if (!res.ok) throw new Error(`Failed to fetch history (${res.status})`);
      return res.json();
    },
  });
}

export async function deleteHistoryItem(id: string): Promise<void> {
  const res = await fetch(`${API_BASE}/api/v1/history/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok) throw new Error(`Failed to delete history item (${res.status})`);
  invalidateCached(HISTORY_LIST_CACHE_PREFIX);
  invalidateCached(`${HISTORY_DETAIL_CACHE_PREFIX}${id}`);
  invalidateInsightsCache();
}

export async function clearHistory(): Promise<ClearHistoryResponse> {
  const res = await fetch(`${API_BASE}/api/v1/history/clear`, {
    method: "POST",
    credentials: "include",
  });
  if (!res.ok) throw new Error(`Failed to clear history (${res.status})`);
  invalidateCached(HISTORY_LIST_CACHE_PREFIX);
  invalidateCached(HISTORY_DETAIL_CACHE_PREFIX);
  invalidateInsightsCache();
  return res.json();
}

File: apps/api/app/api/v1/history.py
"""History CRUD endpoints — all require authentication."""

from __future__ import annotations

import json
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_current_user
from app.db.models import ClassificationEvent, User
from app.db.session import get_db_session
from app.schemas.history import (
    ClearHistoryResponse,
    FeedbackSummary,
    HistoryDetailResponse,
    HistoryItemResponse,
    HistoryListResponse,
)
from app.services import history_service

router = APIRouter()


def _to_item(event: ClassificationEvent) -> HistoryItemResponse:
    return HistoryItemResponse(
        id=UUID(event.id),
        source=event.source,
        subject=event.subject_snippet,
        sender=event.sender,
        final_prediction=event.final_prediction,
        final_risk_score=event.final_risk_score,
        risk_band=event.risk_band,
        personalized=event.personalized,
        saved_at=event.created_at,
    )


def _to_detail(event: ClassificationEvent) -> HistoryDetailResponse:
    reasons: Optional[list[str]] = None
    if event.personalization_reasons:
        try:
            reasons = json.loads(event.personalization_reasons)
        except Exception:
            reasons = None

    feedback_list = [
        FeedbackSummary(
            feedback_label=fb.feedback_label,
            reason=fb.reason,
            created_at=fb.created_at,
        )
        for fb in (event.feedback or [])
    ]

    return HistoryDetailResponse(
        id=UUID(event.id),
        source=event.source,
        subject=event.subject_snippet,
        sender=event.sender,
        final_prediction=event.final_prediction,
        final_risk_score=event.final_risk_score,
        risk_band=event.risk_band,
        personalized=event.personalized,
        saved_at=event.created_at,
        review_state=event.review_state,
        personalization_reasons=reasons,
        agreement_ratio=event.agreement_ratio,
        model_version=event.model_version,
        feedback=feedback_list,
    )


@router.get("/history", response_model=HistoryListResponse)
async def list_history(
    cursor: Optional[str] = Query(default=None),
    limit: int = Query(default=20, ge=1, le=100),
    source: Optional[str] = Query(default=None, pattern="^(manual|gmail)$"),
    verdict: Optional[str] = Query(default=None, pattern="^(spam|not_spam|review)$"),
    q: Optional[str] = Query(default=None, max_length=128),
    user: User = Depends(get_current_user),
) -> HistoryListResponse:
    async with get_db_session() as session:
        if session is None:
            raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database unavailable")
        items, next_cursor = await history_service.list_events(
            session,
            user_id=user.id,
            cursor=cursor,
            limit=limit,
            source_filter=source,
            verdict_filter=verdict,
            query=q,
        )
    return HistoryListResponse(
        items=[_to_item(e) for e in items],
        next_cursor=next_cursor,
    )


@router.get("/history/{history_id}", response_model=HistoryDetailResponse)
async def get_history_item(
    history_id: str,
    user: User = Depends(get_current_user),
) -> HistoryDetailResponse:
    async with get_db_session() as session:
        if session is None:
            raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database unavailable")
        event = await history_service.get_event(session, user_id=user.id, event_id=history_id)
    if event is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="History item not found")
    return _to_detail(event)


@router.delete("/history/{history_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_history_item(
    history_id: str,
    user: User = Depends(get_current_user),
) -> None:
    async with get_db_session() as session:
        if session is None:
            raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database unavailable")
        deleted = await history_service.delete_event(session, user_id=user.id, event_id=history_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="History item not found")


@router.post("/history/clear", response_model=ClearHistoryResponse)
async def clear_history(
    user: User = Depends(get_current_user),
) -> ClearHistoryResponse:
    async with get_db_session() as session:
        if session is None:
            raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database unavailable")
        count = await history_service.clear_events(session, user_id=user.id)
    return ClearHistoryResponse(deleted_count=count)
